content/blog/2020-04-preemption.md

```diff
@@ -0,0 +1,183 @@
+    });
+}
+```
+
+The Rust compiler transforms this code into a state machine. The Tokio runtime
+executes these state machines, multiplexing many tasks on a handful of threads.
+Tokio's scheduler requires that the generated task state machine yields control
+back to the scheduler in order to multiplex tasks. Each `.await` call is an
+opportunity to yield back to the scheduler. In the above example,
+`listener.accept().await` will return a socket if one is pending. If there are
+no pending sockets, control is yielded back to the scheduler.
+
+This system works well in most cases. However, when a system comes under load,
+it is possible for an asynchronous resource to never be "not ready". For
+example, consider an echo server:
+
+```rust
+tokio::spawn(async move {
+    let mut buf = [0; 1024];
+
+    loop {
+        let n = socket.read(&mut buf).await?;
+
+        if n == 0 {
+            break;
+        }
+
+        // Write the data back
+        socket.write(buf[..n]).await?;
+    }
+});
+```
+
+If data is received faster than it can be processed, it is possible that, by the
+time the processing of a data chunk completes, more data has already been
+received. In this case, `.await` will never yield control back to the scheduler,
+other tasks will not be scheduled, resulting in starvation and large latency
+variance.
+
+Currently, the answer to this problem is that the user of Tokio is responsible
+for adding yield points every so often. In practice, very few actually do this
+and end up being vulnerable to this sort of problem.
+
+A common solution to this problem is preemption. OS threads will interrupt
+thread execution every so often in order to ensure fair scheduling of all
+threads. Runtimes that have full control over execution (golang, erlang, ...)
+will also use preemption to ensure fair scheduling of tasks. This is
+accomplished by injecting yield points when compiling code that check if the
+task has been executing for long enough and yielding back to the scheduler.
+Unfortunately, Tokio is not able to use this technique as `async` Rust does not
+inject any sort of yield point.
+
+## Per-task operation budget
+
+Even though Tokio is not able to **preempt**, there is still an opportunity to
+nudge a task to yield back to the scheduler. As of [0.2.14], each Tokio task has
+an operation budget. This budget is reset when the scheduler switches to the
+task. Each Tokio resource (socket, timer, channel, ...) is now aware of this
+budget. As long as the task as budget remaining, the resource operates as it did
+previously. Each asynchronous operation (actions that users must `.await` on)
+decrement the task's budget. Once the task is out of budget, all resources will
+become "not ready" until the task yields back to the scheduler, at which point,
+the budget is reset.
+
+Going back to the echo server example from above. When the task is scheduled, it
+is assigned a budget of 128 operations. When `socket.read(..)` and
```

Comment:
This 128 sticks out a bit. May warrant a sentence of explanation.

Reply:
Say more? I don't think the exact number is important, but it seemed weird to write w/o specifying the number. I could sub it for `n`.

Reply:
Ah, sorry, what I meant was that maybe we should have a sentence about why we chose 128, and how it's an arbitrary choice that will be revisited with more experience?

Reply:
Or maybe even eventually dynamically adjusted using some heuristic?